chain-tx-enclave/enclave-ra/ra-client/src/verifier.rs

```diff
@@ -0,0 +1,226 @@
+use std::{collections::HashSet, fs::read, sync::Arc, time::SystemTime};
+
+use der_parser::oid::Oid;
+use ra_common::{AttestationReport, AttestationReportBody, OID_EXTENSION_ATTESTATION_REPORT};
+use rustls::{
+    internal::pemfile::certs, Certificate, ClientConfig, RootCertStore, ServerCertVerified,
+    ServerCertVerifier, TLSError,
+};
+use sha2::{Digest, Sha256};
+use thiserror::Error;
+use webpki::{
+    DNSNameRef, EndEntityCert, SignatureAlgorithm, TLSServerTrustAnchors, Time, TrustAnchor,
+    ECDSA_P256_SHA256, RSA_PKCS1_2048_8192_SHA256,
+};
+use x509_parser::parse_x509_der;
+
+use crate::EnclaveCertVerifierConfig;
+
+static SUPPORTED_SIG_ALGS: &[&SignatureAlgorithm] = &[&ECDSA_P256_SHA256];
+
+pub struct EnclaveCertVerifier {
+    root_cert_store: RootCertStore,
+    valid_enclave_quote_statuses: HashSet<String>,
+}
+
+impl EnclaveCertVerifier {
+    /// Creates a new instance of enclave certificate verifier
+    pub fn new(config: EnclaveCertVerifierConfig) -> Result<Self, EnclaveCertVerifierError> {
+        let signing_ca_cert_pem = read(config.signing_ca_cert_path.as_ref())?;
+
+        let mut root_cert_store = RootCertStore::empty();
+        root_cert_store
+            .add_pem_file(&mut signing_ca_cert_pem.as_ref())
+            .map_err(|_| EnclaveCertVerifierError::CertificateParsingError)?;
+
+        let mut valid_enclave_quote_statuses =
+            HashSet::with_capacity(config.valid_enclave_quote_statuses.as_ref().len());
+
+        for status in config.valid_enclave_quote_statuses.as_ref() {
+            valid_enclave_quote_statuses.insert(status.clone().into_owned());
+        }
+
+        Ok(Self {
+            root_cert_store,
+            valid_enclave_quote_statuses,
+        })
+    }
+
+    /// Verifies certificate
+    fn verify_cert(&self, certificate: &[u8]) -> Result<(), EnclaveCertVerifierError> {
+        let (_, certificate) = parse_x509_der(certificate)
+            .map_err(|_| EnclaveCertVerifierError::CertificateParsingError)?;
+
+        if certificate
+            .tbs_certificate
+            .validity
+            .time_to_expiration()
+            .is_none()
+        {
+            return Err(EnclaveCertVerifierError::CertificateExpired);
+        }
+
+        let attestation_report_oid = Oid::from(OID_EXTENSION_ATTESTATION_REPORT);
+        let mut attestation_report_received = false;
+
+        let public_key_hash = sha256(
```

Comment:
Possibly. But [this](https://arxiv.org/ftp/arxiv/papers/1801/1801.05863.pdf) paper mentions that we should put hash of public key.

Reply:
I assume it's because, in there, it's meant for any TLS-algo, including RSA

Reply:
for ECC, this will make it simpler + possible to directly use it and test it in client-core

Reply:
Report data is 64 bytes. We can fit public key in report data by stripping off the first byte. But, this won't enable us to directly use it in `client-core` without new `tx-query`. Also, I didn't want to deviate from what is mentioned in the paper.

Reply:
I don't see "bincode" anywhere in the paper, so it deviated anyway -- making it simpler or compatible with existing code is highly desirable

Reply:
Paper does not talk about any kind of binary serialization. I you feel bincode is bad and we shouldn't use it, can you suggest a serialization format that you're comfortable with? Because, to send data over a TCP stream, you need to serialize your data in some format (I chose bincode but I'm okay with any other format you think is more suitable).

Also, I don't see any problem with hashing public keys. Anyway, it doesn't matter and I'll remove it.

Reply:
https://github.com/crypto-com/chain/pull/1311#discussion_r398435690

Reply:
Changed. Now we store public key in report data. Note that report data is 64 bytes and uncompressed public key is 65 bytes, so, we strip off the first byte (it should be `4`) of public key before passing it in report data.